api/handlers/datasources.py

```diff
@@ -159,3 +161,39 @@ async def sync_datasource(
         raise HTTPException(status_code=404, detail="Datasource not found")
     background_tasks.add_task(datasource_repo.sync, account, container)
     return {"message": "Syncing"}
+
+
+@router.get("/api/datasources/{account}/{container}/{file_path}/sas")
+async def generate_sas_token(
+    account: str,
+    container: str,
+    file_path: str,
+    datasources_collection: AgnosticCollection = Depends(datasource_repo.collection),
+    current_user: Optional[dict] = Depends(get_current_user),
+    access_allowed=Depends(check_access),
+):
+    token: str = ""
+    existing_datasource = await datasources_collection.find_one(
+        {
+            "account": account,
+            "container": container,
+        }
+    )
+    if not existing_datasource:
+        raise HTTPException(status_code=404, detail="Datasource not found")
+    if not existing_datasource.get("account_key", None):
+        if not existing_datasource["sasToken"] or access_allowed == "public":
```

Comment:
Looking at it now, I think we should split up the access allowed + no sas token check, just in case there's no sas token and there should be. There could be a case where public access isn't allowed but for some reason the datasource wasn't given an sas token on upload, so we should differentiate between those when telling the front end why there isn't a token so the user can know if there's an error. Maybe if public access allowed we can just return a blank token or null or something which should be fine, and then an error if public access isn't allowed and there's no sas

Reply:
ok, I've added the separation

Reply:
cheers!